React/React hooks.md

```diff
@@ -1,7 +1,68 @@
 ### React hooks
 
-#### Using hook dependency arrays
-Dependency arrays are a way to trigger re-firing of hooks only when it's dependencies change. React will keep the values generated by a hook stored in memory and then run it when a change of a dependency has occured. This applies to all hooks that can have dependencies.
+React Hooks were introduced in version 16.8.0 as Function Component counterpart of Class Component Lifecycle.  
+For more information, see the official [React Hooks API Reference](https://reactjs.org/docs/hooks-reference.html).
+
+#### Hooks flow
+Before we start using hooks in React we need to understand the flow.
+
+Here is a flow diagram that explains the new flow of a Hooks component.
+<br/>
+<br/>
+![hooks-flow](https://raw.githubusercontent.com/donavon/hook-flow/master/hook-flow.png)
+
+> Most important thing here is to notice how "Run Effects" phase is executed last!
+
+Here is the code example that explains this flow.
+
+```jsx
+export const MyComponent: FC = () => {
+  // 1. Run Lazy initializers (i.e. () => 0)
+  const [state, setState] = useState(() => 0); 
+
+  const previousStateRef = useRef();
+
+  useEffect(() => {
+    // 3. Run effect
+    previousStateRef.current = state;
+  }); // no deps array, we want this to be called on every render!
+
+  // 2. Render
+  return (
+    <>
+      Prev: {previousStateRef.current ?? "undefined"}
+      <br />
+      Current: {state}
+      <br />
+      <button onClick={() => setState(1)}>Update</button>
+    </>
+  );
+}
+```
+
+In the "Mount" phase you will see:
+```
+Prev: undefined
+Current: 0
+```
+This is because `useEffect` is called after the first render and also assigning values to `ref` does not trigger re-render, instead, the value is populated and waiting for next update phase.
```

Comment:
```suggestion
This is because `useEffect` is called after the first render and because assigning a value to `ref` does not trigger a re-render - instead, the value is populated and waiting for the next update phase.
```